IPython/extensions/deduperreload/deduperreload.py

```diff
@@ -0,0 +1,477 @@
+import ast
+import contextlib
+import os
+import sys
+import textwrap
+from types import ModuleType
+from typing import TYPE_CHECKING, Any, NamedTuple
+
+from IPython.extensions.deduperreload.deduperreload_patching import (
+    DeduperReloaderPatchingMixin,
+)
+
+if TYPE_CHECKING:
+    from IPython.core.interactiveshell import ExecutionResult
+
+    TDefinitionAst = ast.FunctionDef | ast.AsyncFunctionDef
+
+DefinitionAst = (ast.FunctionDef, ast.AsyncFunctionDef)
+
+
+def get_module_file_name(module: ModuleType | str) -> str:
+    """Returns the module's file path, or the empty string if it's inaccessible"""
+    if (mod := sys.modules.get(module) if isinstance(module, str) else module) is None:
+        return ""
+    return getattr(mod, "__file__", "") or ""
+
+
+def compare_ast(node1: ast.AST | list[ast.AST], node2: ast.AST | list[ast.AST]) -> bool:
+    """Checks if node1 and node2 have identical AST structure/values, apart from some attributes"""
+    if type(node1) is not type(node2):
+        return False
+
+    if isinstance(node1, ast.AST):
+        for k, v in node1.__dict__.items():
+            if k in (
+                "lineno",
+                "end_lineno",
+                "col_offset",
+                "end_col_offset",
+                "ctx",
+                "parent",
+            ):
+                continue
+            if not hasattr(node2, k) or not compare_ast(v, getattr(node2, k)):
+                return False
+        return True
+
+    elif isinstance(node1, list) and isinstance(node2, list):
+        return len(node1) == len(node2) and all(
+            compare_ast(n1, n2) for n1, n2 in zip(node1, node2)
+        )
+    else:
+        return node1 == node2
+
+
+class DependencyNode(NamedTuple):
+    """
+    Each node represents a function.
+    qualified_name: string which represents the namespace/name of the function
+    abstract_syntax_tree: subtree of the overall module which corresponds to this function
+
+    qualified_name is of the structure: (namespace1, namespace2, ..., name)
+
+    For example, foo() in the following would be represented as (A, B, foo):
+
+    class A:
+        class B:
+            def foo():
+                pass
+    """
+
+    qualified_name: tuple[str, ...]
+    abstract_syntax_tree: ast.AST
+
+
+class AutoreloadTree:
+    """
+    Recursive data structure to keep track of reloadable functions/methods. Each object corresponds to a specific scope level.
+    children: classes inside given scope, maps class name to autoreload tree for that class's scope
+    funcs_to_autoreload: list of function names that can be autoreloaded in given scope.
+    new_nested_classes: Classes getting added in new autoreload cycle
+    """
+
+    def __init__(self) -> None:
+        self.children: dict[str, AutoreloadTree] = {}
+        self.defs_to_reload: dict[str, ast.AST] = {}
+        self.defs_to_delete: set[str] = set()
+        self.new_nested_classes: dict[str, ast.AST] = {}
```

Comment:
I think usually you want to the type annotation at class level for proper type checking in subclasses I believe; it used to be that mypy was not able to infer type annotation of attribute defined in __init__. (or emit a `AutoreloadTree.children` attribute might not be set/ None does not have method XXX).

I don't think that's a big problem here, but something to keep in mind depending on what

Reply:
I believe that newer versions of mypy handle this properly now (though I could be wrong)